src/shared/openssl-util.c

```diff
@@ -269,6 +279,177 @@ int string_hashsum(
 
 }
 #  endif
+
+/* We use derived shared secret as volume key for EC keys. The following is the procedure. First we generate
+ * a pair of EC key of the same curve group. Then we use newly generated private key to derive shared secret
+ * with the public key in the token. Finally, we simplely forget the private key, store the public key to
+ * pkcs11-key field of the LUKS2 header and use the derived shared secret as the volume key (we don't apply
+ * any KDF here since LUKS will apply PBKDF on the key).
+ */
+static int pkey_generate_volume_key_ec(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_saved_key,
+                size_t *ret_saved_key_size) {
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+        _cleanup_free_ uint8_t *decrypted_key = NULL;
+        _cleanup_free_ uint8_t *saved_key = NULL;
+        size_t decrypted_key_size, saved_key_size;
+        int nid = NID_undef;
+        int r;
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+        _cleanup_free_ char *curve_name = NULL;
+        size_t len = 0;
+        if (EVP_PKEY_get_group_name(pkey, NULL, 0, &len) != 1 || len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine PKEY group name length");
+        len += 1;
+        curve_name = new(char, len);
+        if (!curve_name)
+                return log_oom();
+
+        if (!EVP_PKEY_get_group_name(pkey, curve_name, len, &len))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get PKEY group name");
+        nid = OBJ_sn2nid(curve_name);
+#else
+        EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+        if (!ec_key)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "PKEY doesn't have EC_KEY associated");
+
+        if (EC_KEY_check_key(ec_key) != 1)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "EC_KEY associated with PKEY is not valid");
+
+        nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+#endif
+
+        r = pkey_generate_ec_key(nid, &pkey_new);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate ec key: %m");
+
+        r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, NULL, &decrypted_key_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine derived shared secret size: %m");
+
+        if (decrypted_key_size == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine derived shared secret size");
+
+        decrypted_key = new(uint8_t, decrypted_key_size);
```

Comment:
this is not really an array for us, it's just "some binary blob", right? hence really should be consider a `void*`, not a `uint8_t*` for us, and hence be allocated with a simple `malloc()`